### PR TITLE
ENTERPRISE-765: introduce quality gates based on intelligent notes

### PR DESCRIPTION
**Pull request by adamtornhill**

### CODE REVIEW

File: src/test/java/org/jenkinsci/plugins/codescene/Domain/DeltaAnalysisRequestTest.java [hunk 1]
```diff
@@ -1,35 +1,129 @@
...
 import static org.junit.Assert.*;
 
 public class DeltaAnalysisRequestTest {
 
     private static final Repository GIT_REPO = new Repository("codescene-ui");
     private static final int COUPLING_THRESHOLD = 65;
 
+    // emulate named parameters to get more calling context
+    private String useBiomarkers = "";
+    private boolean enableUserBiomarkers = false;
```
Comment (jumarko) [inline src/test/java/org/jenkinsci/plugins/codescene/Domain/DeltaAnalysisRequestTest.java]: should be `enableUseBiomarkers`?

File: src/test/java/org/jenkinsci/plugins/codescene/Domain/DeltaAnalysisRequestTest.java [hunk 2]
```diff
@@ -1,35 +1,129 @@
...
+    private static boolean enableFailedGoalGate = false;
+    private static boolean enableCodeHealthGate = false;
```
Comment (jumarko) [inline src/test/java/org/jenkinsci/plugins/codescene/Domain/DeltaAnalysisRequestTest.java]: I don't like this workaround for "named parameters".
The fields should be removed and userConfigFrom's params values just passed without assigning them to these fields.
If you really want to do it, then Builder pattern may be a better solution.

File: src/main/java/org/jenkinsci/plugins/codescene/Domain/QualityGates.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
+package org.jenkinsci.plugins.codescene.Domain;
+
+import javax.json.JsonObject;
+
+public class QualityGates {
+    private boolean goalHasFailed = false;
+    private boolean codeHealthDeclined = false;
+    private boolean gatesEnabled = false;
```
Comment (jumarko) [inline src/main/java/org/jenkinsci/plugins/codescene/Domain/QualityGates.java]: I'm not sure if there's any benefit in having redundant `false` initializers.

File: src/main/java/org/jenkinsci/plugins/codescene/CodeSceneBuilder.java [hunk 4]
```diff
@@ -380,6 +400,40 @@ private UsernamePasswordCredentials lookupCredentials(String credentialId) {
...
+            Result newResult = Result.UNSTABLE;
+
+            Result result = build.getResult();
+            if (result != null) {
+                build.setResult(result.combine(newResult));
+            } else {
+                build.setResult(newResult);
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (adamtornhill) [reply src/main/java/org/jenkinsci/plugins/codescene/CodeSceneBuilder.java]: While they are related, they are *not* the same. One is based on contextual information and, hence, most important and relevant to everyone. A code health decline on the other hand is orthogonal to the goals; for example, a goal can fail *without* a decline in code health (think of a planned refactoring that never happens).
So I want to keep this a separate configs.

File: src/main/java/org/jenkinsci/plugins/codescene/CodeSceneBuildActionEntry.java [hunk 5]
```diff
@@ -74,4 +74,18 @@ public int getRiskThreshold() {
     public boolean getHitsRiskThreshold() {
         return risk.getValue() >= riskThreshold;
     }
+
+    public QualityGates gates() { return this.qualityGatesState; }
+
+    public boolean getGoalHasFailed() {
+        return qualityGatesState.goalHasFailed();
```
[the remark this answers is not in the document]
Comment (adamtornhill) [reply src/main/java/org/jenkinsci/plugins/codescene/CodeSceneBuildActionEntry.java]: Yes, needs to be fixed. How do I reproduce it?

File: src/main/java/org/jenkinsci/plugins/codescene/CodeSceneBuilder.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jumarko) [reply src/main/java/org/jenkinsci/plugins/codescene/CodeSceneBuilder.java]: I can see that the gates themselves server a different purpose from a user's perspective.
It's just that the implementations of aforementioned methods look almost identical so I was wondering if the duplicated code can be abstracted away.

File: src/main/java/org/jenkinsci/plugins/codescene/CodeSceneBuildActionEntry.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jumarko) [reply src/main/java/org/jenkinsci/plugins/codescene/CodeSceneBuildActionEntry.java]: I think, I got it the first time I updated the plugin in jenkins and then just ran an existing job (without even looking into configuration or doing anything else).

**Comment by jumarko** [comment]
[Introduce CI/CD quality gates based on intelligent notes.](https://trello.com/c/v0XdSwfv/765-introduce-ci-cd-quality-gates-based-on-intelligent-notes)

File: src/main/java/org/jenkinsci/plugins/codescene/CodeSceneBuilder.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jumarko) [reply src/main/java/org/jenkinsci/plugins/codescene/CodeSceneBuilder.java]: fixed by extracting common code to a new private method.

File: src/main/java/org/jenkinsci/plugins/codescene/Domain/DeltaAnalysisRequest.java [hunk 6]
```diff
@@ -7,19 +7,26 @@
...
-                                boolean useBiomarkers) {
+    public DeltaAnalysisRequest(final Commits commits, final Configuration userConfig) {
         final JsonArray cs = serialize(commits);
 
         JsonObjectBuilder b = Json.createObjectBuilder();
         b.add("commits", cs);
-        b.add("repository", gitRepository.value());
-        b.add("coupling_threshold_percent", couplingThresholdPercent);
-        b.add("use_biomarkers", useBiomarkers);
+        b.add("repository", userConfig.gitRepisitoryToAnalyze().value());
```
[the remark this answers is not in the document]
Comment (jumarko) [reply src/main/java/org/jenkinsci/plugins/codescene/Domain/DeltaAnalysisRequest.java]: fixed.

File: src/test/java/org/jenkinsci/plugins/codescene/Domain/DeltaAnalysisRequestTest.java (hunk 1 above)
Comment (jumarko) [reply src/test/java/org/jenkinsci/plugins/codescene/Domain/DeltaAnalysisRequestTest.java]: fixed.

File: src/test/java/org/jenkinsci/plugins/codescene/Domain/DeltaAnalysisRequestTest.java (hunk 2 above)
Comment (jumarko) [reply src/test/java/org/jenkinsci/plugins/codescene/Domain/DeltaAnalysisRequestTest.java]: Named params removed.

File: src/main/java/org/jenkinsci/plugins/codescene/Domain/QualityGates.java (hunk 3 above)
Comment (jumarko) [reply src/main/java/org/jenkinsci/plugins/codescene/Domain/QualityGates.java]: removed.

**Pull request merged**